runtime/lua/vim/shared.lua

```diff
@@ -396,6 +396,37 @@ function vim.tbl_deep_extend(behavior, ...)
   return tbl_extend(behavior, true, ...)
 end
 
+--- Traverses any tree-like object defined by `key`,
+--- calling the provided `visitor` on each descendant.
+---
+--- Example:
+---
+--- ```lua
+--- local comments = {}
+--- vim.traverse(
+---   language_tree,
+---   function(tree) return tree:children() end,
+---   function(tree)
+---     if tree:lang() == "comment" then
+---       table.insert(comments, tree)
+---     end
+---   end
+--- )
+--- ```
+---
+---@param tree any The object to traverse.
+---@param key string|fun(subtree: any): any[] Define the "children" to visit.
+---@param visitor fun(subtree: any): boolean|nil Function to call on each descendant.
+---       Return true to skip the rest of `subtree`.
+function vim.traverse(tree, key, visitor)
+  if not visitor(tree) then
+    local children = type(key) == 'function' and key(tree) or tree[key]
```

Comment:
In Lua any value is a valid index, including functions, so this check will have an unexpected effect if a tree uses functions as its keys. I know this is a very weird use-case, but I think it is best avoided.

Reply:
Are you saying a user might want to pass in a `type(key) == 'function'` but get a `tree[key]`? I think the user can accomplish that with `vim.traverse(tree, function(tree) return tree[key] end, visitor)` instead of `vim.traverse(tree, key, visitor)`.